shell/gpu/gpu_surface_metal.mm

```diff
@@ -90,16 +113,33 @@
 
     canvas->flush();
 
-    GrMTLHandle drawable = next_drawable_;
-    if (!drawable) {
-      FML_DLOG(ERROR) << "Unable to obtain a metal drawable.";
-      return false;
+    uintptr_t texture = reinterpret_cast<uintptr_t>(drawable.get().texture);
+    for (auto& entry : damage_) {
+      if (entry.first != texture) {
+        // Accumulate damage for other framebuffers
+        for (const auto& rect : surface_frame.submit_info().frame_damage) {
+          entry.second.join(rect);
+        }
+      }
     }
+    // Reset accumulated damage for current framebuffer
+    damage_[texture] = SkIRect::MakeEmpty();
 
     return delegate_->PresentDrawable(drawable);
   };
 
-  return std::make_unique<SurfaceFrame>(std::move(surface), true, submit_callback);
+  SurfaceFrame::FramebufferInfo framebuffer_info;
+  framebuffer_info.supports_readback = true;
+
+  // Provide accumulated damage to rasterizer (area in current framebuffer that lags behind
+  // front buffer)
+  uintptr_t texture = reinterpret_cast<uintptr_t>(drawable.get().texture);
+  auto i = damage_.find(texture);
+  if (i != damage_.end()) {
+    framebuffer_info.existing_damage.push_back(i->second);
+  }
```

Comment:
`damage_` is a map, should it perhaps be a circular linked list so that you can run through all buffers back until the last time you used this one?

Reply:
Not sure I see the benefit. Damage is accumulated per texture. Which texture is being used depends on what CAMetalLayer gives us.